File: leetcode/236.lowest-common-ancestor-of-a-binary-tree.py
# @leet start
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None


class Solution:
    def lowestCommonAncestor(
        self, root: TreeNode, p: TreeNode, q: TreeNode
    ) -> TreeNode:
        return self.helper(root, p, q)

    def helper(
        self, root: TreeNode | None, p: TreeNode, q: TreeNode
    ) -> TreeNode | None:
        if not root or root in (p, q):
            return root

        left = self.helper(root.left, p, q)
        right = self.helper(root.right, p, q)

        if left and right:
            return root

        return left or right


# @leet end

File: leetcode/236.lowest-common-ancestor-of-a-binary-tree.js
// @leet start
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @param {TreeNode} p
 * @param {TreeNode} q
 * @return {TreeNode}
 */
var lowestCommonAncestor = function (root, p, q) {
  const helper = (node) => {
    if (node === null || node === p || node === q) {
      return node;
    }

    const left = helper(node.left);
    const right = helper(node.right);

    return left !== null && right !== null ? node : (left ?? right);
  };

  return helper(root);
};
// @leet end
